### PR TITLE
[REF] get_payment_limits, get_summary_for_cashregister

### PR DESCRIPTION
**Pull request by KolushovAlexandr**

### CODE REVIEW

File: pos_debt_notebook/static/src/js/pos.js [hunk 1]
```diff
@@ -203,11 +203,41 @@ odoo.define('pos_debt_notebook.pos', function (require) {
             }
             return data;
         },
+        get_payment_limits: function(cashregister, arg){
+            var self = this;
+            var arg = arg || '';
+            var vals = {};
+            if (cashregister.journal.debt){
+                var partner_balance = this.pos.get_client().debts[cashregister.journal.id].balance,
+                    sum_pl = this.get_summary_for_cashregister(cashregister),
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: it is not used

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/static/src/js/pos.js [hunk 2]
```diff
@@ -209,11 +209,40 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
+                var partner_balance = this.pos.get_client().debts[cashregister.journal.id].balance,
+                    category_list = cashregister.journal.category_ids;
+                var sum_prod = category_list.length
+                    ? this.get_summary_for_categories(category_list)
+                    : this.get_total_with_tax();
+                if (arg === 'debt_limit' || arg === 'all'){
+                    vals.debt_limit = partner_balance + cashregister.journal.debt_limit;
+                }
+                if (category_list.length && (arg === 'products_restriction' || arg === 'all')){
+                    vals.products_restriction = category_list.length
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: category_list.length is always truthy here

**Review by yelizariev — CHANGES_REQUESTED** [review]
Also, add documentation for get_payment_limits (comment right in the beginning of the method

File: pos_debt_notebook/static/src/js/pos.js [hunk 3]
```diff
@@ -240,24 +273,12 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
-                    var sum_prod = this.get_summary_for_categories(category_list);
-                    amount = Math.max(Math.min(
-                        sum_prod - sum_pl,
-                        amount,
-                        journal.debt_limit - sum_pl + partner_balance), 0);
                 } else {
-                    amount = Math.max(Math.min(
-                        amount,
-                        journal.debt_limit - sum_pl + partner_balance), 0);
+                    amount = Math.max(Math.min.apply(null, _.values(limits)) - this.get_summary_for_cashregister(cashregister), 0);
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: You can compute limits only if it's needed

File: pos_debt_notebook/static/src/js/pos.js [hunk 4]
```diff
@@ -240,24 +273,12 @@ odoo.define('pos_debt_notebook.pos', function (require) {
                 pos: this.pos
             });
             if (cashregister.journal.debt && this.get_client()){
-                var category_list = cashregister.journal.category_ids;
-                var partner_balance = this.pos.get_client().debts[journal.id].balance;
+                var limits = this.get_payment_limits(cashregister, 'all');
                 var amount = this.get_due();
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: unused computing

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/models.py [hunk 5]
```diff
@@ -465,6 +465,7 @@ def _process_order(self, pos_order):
                                        'note': product_list,
                                        })
                 payment[2]['amount'] = 0
+        pos_order['amount_return'] -= pos_order['amount_via_discount']
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: amount_via_discount may not be presented. Use  pos_order.get('amount_via_discount', 0)

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/static/src/js/pos.js [hunk 6]
```diff
@@ -465,15 +473,11 @@ odoo.define('pos_debt_notebook.pos', function (require) {
             var order = this.pos.get_order(),
             flag = false;
             if (this.pos.get_client()){
-                var paymentlines = order.get_paymentlines(),
-                debts = this.pos.get_client().debts;
-                _.each(paymentlines, function(pl){
-                    var cr = pl.cashregister;
+                _.each(this.get_used_debt_cashregisters(), function(cr){
                     if (cr.journal.debt) {
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Always true here

File: pos_debt_notebook/static/src/js/pos.js [hunk 7]
```diff
@@ -483,14 +487,15 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
-                paymentlines = order.get_paymentlines(),
                 flag = false;
-            for (var i = 0; i < paymentlines.length; i++) {
-                var journal = paymentlines[i].cashregister.journal;
-                if (order.get_change(paymentlines[i]) > 0 && journal.debt && !journal.pos_cash_out){
+            var cashregisters = this.get_used_debt_cashregisters();
+            _.each(cashregisters, function(cr){
+                var sum_pl = order.get_summary_for_cashregister(cr);
+                var limits = order.get_payment_limits(cr, 'cash_out');
+                if (sum_pl > limits.cash_out){
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: You must check that there is the restriction for cash_out. Use hasOwnProperty or something

File: pos_debt_notebook/static/src/js/pos.js [hunk 8]
```diff
@@ -426,17 +445,12 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
+            var cashregisters = this.get_used_debt_cashregisters();
+            cashregisters = _.filter(cashregisters, function(cr){
+                return cr.journal.category_ids.length > 0;
             });
             var violations = [];
-            if (paymentlines_with_restrictions) {
-                var cashregisters = _.uniq(_.map(paymentlines_with_restrictions, function(pcr){
-                    return pcr.cashregister;
-                }));
+            if (cashregisters) {
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: No need to check it if you use ``_.each``

File: pos_debt_notebook/static/src/js/pos.js [hunk 9]
```diff
@@ -448,15 +462,9 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
-            //summary paid by each journal
-            sum_pl = order.get_summary_for_cashregister(cr);
-            sum_pl = round_pr(sum_pl, this.pos.currency.rounding);
-            //summary allowed to pay
-            sum_prod = order.get_summary_for_categories(journal.category_ids);
-            sum_prod = round_pr(sum_prod, this.pos.currency.rounding);
-            if (sum_pl > sum_prod) {
+            var sum_pl = round_pr(order.get_summary_for_cashregister(cr), this.pos.currency.rounding);
+            var limits = order.get_payment_limits(cr, 'products_restriction');
+            if (sum_pl > limits.products_restriction) {
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Add check limits has products_restriction

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/models.py (hunk 5 above)
Comment (KolushovAlexandr) [reply pos_debt_notebook/models.py]: done

File: pos_debt_notebook/static/src/js/pos.js [hunk 10]
```diff
@@ -448,15 +462,9 @@ odoo.define('pos_debt_notebook.pos', function (require) {
         restricted_products_check: function(cr){
             var order = this.pos.get_order();
             var journal = cr.journal;
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Journal is not used

File: pos_debt_notebook/static/src/js/pos.js [hunk 11]
```diff
@@ -465,15 +473,11 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
-                _.each(paymentlines, function(pl){
-                    var cr = pl.cashregister;
+                _.each(this.get_used_debt_cashregisters(), function(cr){
                     if (cr.journal.debt) {
-                        var debt_limit = cr.journal.debt_limit;
+                        var limits = order.get_payment_limits(cr, 'debt_limit');
                         var sum_pl = order.get_summary_for_cashregister(cr);
-                        var balance = pl.order.get_client().debts[cr.journal.id].balance;
-                        if (sum_pl > debt_limit + balance) {
+                        if (sum_pl > limits.debt_limit) {
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Add a comment why no need to check that  debt_limit is in limits (because it's always there)

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/static/src/js/pos.js [hunk 12]
```diff
@@ -483,14 +487,15 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
-                paymentlines = order.get_paymentlines(),
                 flag = false;
-            for (var i = 0; i < paymentlines.length; i++) {
-                var journal = paymentlines[i].cashregister.journal;
-                if (order.get_change(paymentlines[i]) > 0 && journal.debt && !journal.pos_cash_out){
+            var cashregisters = this.get_used_debt_cashregisters();
+            _.each(cashregisters, function(cr){
+                var sum_pl = order.get_summary_for_cashregister(cr);
+                var limits = order.get_payment_limits(cr, 'cash_out');
+                if (limits.cash_out && sum_pl > limits.cash_out){
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: It's not the same. Somehow limits.cash_out may be 0, but you still need to check condition  sum_pl > limits.cash_out

**Review by yelizariev — CHANGES_REQUESTED** [review]

**Review by yelizariev — APPROVED** [review]

**Pull request merged**